codalab/apps/authenz/models.py

```diff
@@ -32,4 +41,57 @@ class ClUser(auth_models.AbstractUser):
     rabbitmq_username = models.CharField(max_length=36, null=True, blank=True)
     rabbitmq_password = models.CharField(max_length=36, null=True, blank=True)
 
+    def save(self, *args, **kwargs):
+        if self.newsletter_opt_in and self.email and self.is_active:
+            data = {
+                "email_address": self.email,
+                "status": "subscribed",
+            }
+
+            user_hash = hashlib.md5(str.lower(self.email.encode()))
+
+            # Try to update the user before creating a new user
+            update_user = requests.patch(
+                settings.MAILCHIMP_MEMBERS_ENDPOINT_NEWSLETTER + '/' + user_hash.hexdigest(),
+                auth=("", settings.MAILCHIMP_API_KEY),
+                data=json.dumps(data)
+            )
+
+            # If no user is found in mailchimp, create a new mailing list user
+            if not update_user.ok:
+                requests.post(
+                    settings.MAILCHIMP_MEMBERS_ENDPOINT_NEWSLETTER,
+                    auth=("", settings.MAILCHIMP_API_KEY),
+                    data=json.dumps(data)
+                )
+
+            # If there is no NewsletterUser for this user, create one
+            if not NewsletterUser.objects.filter(email=self.email).exists():
+                NewsletterUser.objects.create(email=self.email)
+
+        elif not self.email and self.newsletter_opt_in:
+            self.newsletter_opt_in = False
+            raise ValidationError('You must have an email to receive the Codalab newsletter')
+
+        elif not self.newsletter_opt_in and self.email:
+            data = {
+                "status": "unsubscribed",
+            }
+
+            user_hash = hashlib.md5(str.lower(self.email.encode()))
+
+            # Update user in Mailchimp to Unsubscribed
+            requests.patch(
+                settings.MAILCHIMP_MEMBERS_ENDPOINT_NEWSLETTER + '/' + user_hash.hexdigest(),
+                auth=("", settings.MAILCHIMP_API_KEY),
+                data=json.dumps(data)
+            )
+
+            # Remove the user from the NewsletterUser model
+            if NewsletterUser.objects.filter(email=self.email).exists():
```

Comment:
Should be able to do
```python
NewsletterUser.objects.filter(email=self.email).delete()
```
which will delete the objects the queryset returns, if any exist, and do nothing if the qs is empty.

Reply:
Changed the way the logic works to unsubscribe but not delete NewsletterSubscriptions